python/cudf/cudf/core/buffer.py

```diff
@@ -201,6 +311,106 @@ def __getitem__(self, key: slice) -> Buffer:
             data=self.ptr + start, size=stop - start, owner=self.owner
         )
 
+    def _is_cai_zero_copied(self):
+        """
+        Returns a flag, that indicates if the Buffer has been zero-copied.
+        """
+        return self._zero_copied
+
+    def _update_ref(self):
+        """
+        Generate the new proxy reference.
+        """
+        if (self._ptr, self._size) not in Buffer._refs:
+            Buffer._refs[(self._ptr, self._size)] = BufferWeakref(
+                self._ptr, self._size
+            )
+        self._proxy_ref = Buffer._refs[(self._ptr, self._size)]
+
+    def get_ref(self):
+        """
+        Returns the proxy reference.
+        """
+        if self._proxy_ref is None:
+            self._update_ref()
+        return self._proxy_ref
+
+    def has_a_weakref(self):
+        """
+        Checks if the Buffer has a weak-reference.
+        """
+        weakref_count = weakref.getweakrefcount(self.get_ref())
+
+        if weakref_count == 1:
+            # When the weakref_count is 1, it could be a possibility
+            # that a copied Buffer got destroyed and hence this
+            # method should return False in that case as there is only
+            # one Buffer pointing to the device memory.
+            return (
+                weakref.getweakrefs(self.get_ref())[0]() is not self.get_ref()
+            )
+        else:
+            return weakref_count > 0
+
+    def _get_weakref(self):
+        """
+        Returns a weak-reference for the Buffer.
+        """
+        return weakref.ref(self.get_ref(), custom_weakref_callback)
+
+    def copy(self, deep: bool = True):
+        """
+        Return a copy of Buffer.
+
+        Parameters
+        ----------
+        deep : bool, default True
+            If True, returns a deep-copy of the underlying Buffer data.
+            If False, returns a shallow-copy of the Buffer pointing to
+            the same underlying data.
+
+        Returns
+        -------
+        Buffer
+        """
```

Comment:
Yes, the behaviour for the `copy()` method itself is probably not explicitly mentioned in the proposals. But what we currently ended up implementing is: `deep=True` -> actual physical copy, `deep=False` -> view but behaves as copy through CoW. 
See https://github.com/pandas-dev/pandas/blob/3020b8df58927ef0246091d325debce7097ace60/pandas/tests/copy_view/test_methods.py#L11-L47 for the tests covering this.

Previously, we called this `deep=False` a "shallow" copy", i.e you return a new object (a new DataFrame container object in this case) but it still points to the same data (the same block / numpy array in our case). But because this is actually the same data, that also meant that mutating the result would update the parent etc. 
With CoW enabled, it's not longer possible to have a shallow copy in the exact same meaning as before (I am not sure if we should keep using that term to avoid confusion with the previous behaviour, but I also don't directly have a good alternative/ "copy that is actually a view but behaves as copy through Copy-on-Write mechanism" is not an easy term to use repeatedly ;))

Reply:
Thanks. So `deep=True` is an "eager" copy and `deep=False` is a "lazy" copy. Those might be shorter ways of describing things, though with no less baggage associated with the terms.